text/0028-bee-storage/0022-bee-storage.md

```diff
@@ -0,0 +1,214 @@
+
+# Summary
+
+An abstraction for the persistent module which will be exclusively used by the Tangle module
+for CRUD operations.
+[https://www.sqlshack.com/crud-operations-in-sql-server/]
+
+# Motivation
+
+The Tangle is a DAG where each vertex's data is a transaction,
+this data should of course be persisted, it is better to have this logic
+and functionality grouped into one separate module that exposes a clear API
+rather than have it in several places, this will also allow for implementing different database backends
+that will implement the storage API
+
+Whenever a bee module requires updating a transaction, it will do so by interacting with the Tangle module
+which will internally use this storage module
+
+use cases:
+
+1.When a new transaction arrives to the node and is not in the database, we would want to insert it as a new record
+2.When a transaction is required we would want to retrieve it from the database
+3.When we want to update a transaction as being solid
+4.When we want to delete a transaction during the local snapshot process
+
+
+# Detailed design
+
+```rust
+
+use std::{collections::HashMap, collections::HashSet, rc::Rc};
+
+/// A transaction hash. To be replaced later with whatever implementation is required.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct TxHash(u64);
+impl TxHash {
+    pub fn is_genesis(&self) -> bool {
+        unimplemented!()
+    }
+}
+
+/// A transaction address. To be replaced later with whatever implementation is required.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct TxAddress(u64);
+
+/// [https://github.com/iotaledger/bee-rfcs/pull/20]
+pub struct Tx {
+    hash: TxHash,
+    trunk: TxHash,
+    branch: TxHash,
+    body: (),
+}
+
+/// A milestone hash. To be replaced later with whatever implementation is required.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct MilestoneHash(u64);
+impl MilestoneHash {
+    //Returns true if the milestone is the genesis of the Tangle
+    pub fn is_genesis(&self) -> bool {
+        unimplemented!()
+    }
+}
+
+pub struct Milestone {
+    hash: MilestoneHash,
+    index: u32,
+}
+
+pub enum StorageError {
+    TransactionAlreadyExist,
+    TransactionNotFound,
+    UnknownError,
+    //...
+}
+
+pub enum ConnectionError {
+    InvalidUrl,
+    UnknownError,
+    //...
+}
+
+type HashesToApprovers = HashMap<TxHash, Vec<TxHash>>;
+type MissingHashesToRCApprovers = HashMap<TxHash, Vec<Rc<TxHash>>>;
+//This is a mapping between an iota address and it's balance change
+//practically, a map for total balance change over an addresses will be collected
+//per milestone (snapshot_index), when we no longer have milestones, we will have to find
+//another way to decide on a check point where to store an address's delta if we want to snapshot
+type StateDeltaMap = HashMap<TxAddress, i64>;
+
+use async_trait::async_trait;
+
+#[async_trait]
+pub trait Connection<Conn> {
```

Comment:
Could this be associate type instead? Type name (Conn) should also rename I guess.

```suggestion
pub trait Connection {
    type Conn;
```